Mark some debugging FunInd commands as internal and add a note on Equations. (#17004)
coq/coq: doc/sphinx/using/libraries/funind.rst

Functional induction
====================

.. note::

>    Most features of ``FunInd`` are also available in ``Equations``, which

Reviewer: ```suggestion
   Most features of ``FunInd`` are also available in the `Equations`` library, which
```

   is available in the Coq Platform. If ``Equations`` suits your needs,
   we recommend you use it in new code instead of ``FunInd``. ``FunInd`` is
   not deprecated and not planned for removal for now because some features
   are not yet available in ``Equations`` (induction principle generation

>    like :cmd:`Function` does).

Reviewer 2: this last precision seems unclear to me. A (imperfect) suggestion:
(:cmd:`Function` and :cmd:`Equations` generate incompatible induction principles).

Reviewer 2: They are better imho in the sense that they follow the user's intent (wrt the order of patterns and the wildcards), but I am not sure it matters here. Does it?

Reviewer 2: But yes incompatibility is the main (only?) reason for not deprecating Function yet. Porting a proof from the latter to the former needs 1) reordering cases in proofs 2) duplicating cases in proofs.

>    Most features of the ``FunInd`` plugin are also available in the
>    ``Equations`` plugin, which is provided in the Coq Platform. For
>    new projects / new code, it is recommended using ``Equations`` instead
>    of ``FunInd``. ``FunInd`` is not deprecated and not planned for removal
>    yet because porting code from ``FunInd`` to ``Equations`` is difficult
>    (incompatibilities in the generated induction principles).

Reviewer: Can we do a little better on this explanation?

```suggestion
   We recommend using the ``Equations`` plugin for new projects and new code
   instead of ``FunInd`` because <EXPLAIN WHY>.
   Most features of the ``FunInd`` plugin are also available in the
   ``Equations`` plugin, which is provided in the Coq Platform.
   ``FunInd`` is not deprecated and not planned for removal
   yet because porting code from ``FunInd`` to ``Equations`` is difficult
   (incompatibilities in the generated induction principles).
```

Equations is the replacement for FunInd?  Could say that more explicitly.
When would you still use FunInd for new code (what features does it have that Equations does not)?

Less important: Is it relevant to users whether something is a plugin or a library?  Is there any important difference?  If not, I would go with "library" throughout the documentation.

Reviewer 2: Equations subsumes Function in almost every aspect: it supports dependent types whereas Function does not. It supports all ways of defining well founded recursive functions that Function supports. The only thing aspect I am aware of where it does not subsume Function is that the functional induction principles generated by Equations are less user-friendly.

Reviewer 2: So yes Equation is a replacement for Function IIUC.

Author: > Less important: Is it relevant to users whether something is a plugin or a library? Is there any important difference? If not, I would go with "library" throughout the documentation.

I don't know. I tend to make the distinction. What does @palmskog think?

Reviewer 3: I think the distinction between "library" and "plugin" has become less clear over time but is still valuable, for three key reasons:
- nearly all plugins add something to the vernacular language that could not be added otherwise (modulo MetaCoq), so users can know by the word "plugin" that they get new top-level commands or tactics.
- plugins can do almost anything to Coq, so they can't be fully trusted - when you add a plugin (as opposed to a library) this is a useful indicator that that the trusted base may increase, which could be mitigated by `coqchk` or similar
- ML plugins that are not in Coq's CI will bitrot quickly, which users likely will want to know before they spend time to adopt in their project

Author: I pushed a new version.

Functional induction
====================

.. note::

>    The functional induction (FunInd) plugin is considered legacy. For new

Reviewer: ```suggestion
   The functional induction (FunInd) plugin is legacy functionality. For new
```

>    code and new projects, the recommended replacement is `Equations
>    <http://mattam82.github.io/Coq-Equations/>`_, a more powerful plugin
>    available through the Coq Platform, which provides most features from
>    FunInd as well as several others. Refer to the `Equations documentation

Reviewer: ```suggestion
   code and new projects, the recommended replacement is `Equations
   <http://mattam82.github.io/Coq-Equations/>`_, a more powerful plugin
   that provides most of FunInd's features and more.  It can be installed
   through the Coq Platform. Refer to the `Equations documentation
```

Hmm, you like the passive voice better than "we recommend"?  In academic papers, one would avoid the first person, but among friends it's nice to use the first person (plural).

Author: It's a writing habit, I guess, but I don't mind using "we".

   <https://raw.githubusercontent.com/mattam82/Coq-Equations/master/doc/equations.pdf>`_

>    to learn more. FunInd is not deprecated and not planned for removal
>    yet because porting code from FunInd to Equations can be difficult

Reviewer: ```suggestion
   to learn more. FunInd is not deprecated and has not been removed
   yet because porting code from FunInd to Equations can be difficult
```

Author: I think I will keep the previous version, because with your suggestion, it could alert users that it could be removed at any point in the future, which is not the case.

Reviewer: Sounds like we have no specific plan (eg a target date, ways to make the migration easier) and may never have one.

One could also remove the "yet", but enough with the wordsmithing.

Author: There is no specific plan, but we know that we want to get rid of this code eventually.
